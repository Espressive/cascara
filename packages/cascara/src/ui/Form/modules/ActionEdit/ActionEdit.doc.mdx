---
title: Action Edit
propTable: ActionEdit.js
type: 'module'
public: false
---

A special action module, fully controled by **Cascara**. Use it when you want the **From** to be editable.
In contrast with the **button** module, this module renders up to 2 buttons at the same time, depending on the **edit** state of the parent context:

1. In **display** mode, it will render a single button with the text you pass as `editLabel` prop, clicking that button will make the **edit** mode kick-in
2. In **edit** mode, it will render a button for exiting **edit** mode (AKA cancel) and one for accepting the changes made during **edit** mode (AKA submit).

All these three buttons will cause your Table or Form to emit the following `onAction` events:

1. **edit.start** when entering in **edit** mode
2. **edit.cancel** when exiting the **edit** mode via the **cancel** button
3. **edit.save** when exiting the **edit** mode via the **save** button

## Using the module

Additional to the props specified in the aside table, you need to pass the **module** prop, with its value set to **edit**.

### Sandbox

Feel free to play around with the following code:

```jsx
<ModuleSandbox>
  <ActionEdit cancelLabel='Cancelar' editLabel='Editar' saveLabel='Guardar' />
</ModuleSandbox>
```
