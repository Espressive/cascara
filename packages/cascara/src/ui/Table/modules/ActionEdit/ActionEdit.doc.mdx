---
title: ActionEdit
propTable: ActionEdit.js
type: 'module'
status: 'Draft'
public: false
---

```
<ActionEdit />
```
