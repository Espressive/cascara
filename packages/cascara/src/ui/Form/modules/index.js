// Form-specific Action Modules
import ActionEdit from './ActionEdit';
import FormRow from './FormRow';

export const formActionModules = {
  edit: ActionEdit,
};

export const formModules = {
  row: FormRow,
};
