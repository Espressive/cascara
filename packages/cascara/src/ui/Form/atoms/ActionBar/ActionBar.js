import React from 'react';
import pt from 'prop-types';

// external prop types
import { propTypes as actionButtonPT } from '../../../../modules/ActionButton';
import { propTypes as actionEditPT } from '../../modules/ActionEdit';

import styles from '../../Form.module.scss';
import ErrorBoundary from '../../../../shared/ErrorBoundary';

const propTypes = {
  actions: pt.arrayOf(
    pt.oneOfType([pt.shape(actionButtonPT), pt.shape(actionEditPT)])
  ),
};

const ActionBar = ({ actions }) => {
  return (
    <ErrorBoundary>
      <div className={styles.FormActionBar}>{actions}</div>
    </ErrorBoundary>
  );
};

ActionBar.propTypes = propTypes;

export default ActionBar;
