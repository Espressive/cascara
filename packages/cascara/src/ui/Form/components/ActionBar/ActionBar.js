import React from 'react';
import pt from 'prop-types';

// external prop types
import { propTypes as actionButtonPT } from '../../../../modules/ActionButton';
import { propTypes as actionEditPT } from '../../modules/ActionEdit';

import styles from '../../Form.module.scss';
import { Boundaries } from '../../../../system-components';

import classnames from 'classnames/bind';

const propTypes = {
  actions: pt.arrayOf(
    pt.oneOfType([pt.shape(actionButtonPT), pt.shape(actionEditPT)])
  ),
  className: pt.string,
};

const cx = classnames.bind(styles);

const ActionBar = ({ actions, className }) => {
  return (
    <Boundaries>
      <div className={cx(className, 'FormActionBar')}>{actions}</div>
    </Boundaries>
  );
};

ActionBar.propTypes = propTypes;

export default ActionBar;
