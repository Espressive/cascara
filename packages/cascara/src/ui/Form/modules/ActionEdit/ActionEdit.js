import React, { useContext } from 'react';
import pt from 'prop-types';

import { ModuleContext } from '../../../../modules/context';
import { Button } from 'reakit';
import { Icon } from 'semantic-ui-react';

const propTypes = {
  cancelLabel: pt.string,
  dataTestIDs: pt.shape({
    cancel: pt.string,
    edit: pt.string,
    save: pt.string,
  }),
  editLabel: pt.string,
  saveLabel: pt.string,
};

const ActionEdit = ({ dataTestIDs, editLabel = 'Edit' }) => {
  const {
    isEditing,
    enterEditMode,
    exitEditMode,
    formMethods,
    data,
    onAction,
  } = useContext(ModuleContext);
  const { handleSubmit, formState, reset } = formMethods;
  const { isDirty, isSubmitting } = formState;

  /**
   * this seems like ugly, we need to find a better way
   * to ease testing.. */
  let cancelTestId = {};
  let editTestId = {};
  let saveTestId = {};

  if (typeof dataTestIDs === 'object') {
    cancelTestId['data-testid'] = dataTestIDs['cancel'];
    editTestId['data-testid'] = dataTestIDs['edit'];
    saveTestId['data-testid'] = dataTestIDs['save'];
  }

  const handleReset = () => {
    onAction(
      // fake target
      {
        name: 'edit.cancel',
      },
      {
        ...data,
      }
    );

    exitEditMode();
  };

  const handleCancel = () => {
    isDirty
      ? // eslint-disable-next-line no-restricted-globals
        confirm('Abandon unsaved changes?') && handleReset()
      : handleReset();
  };

  const handleEdit = () => {
    /**
     * FDS-91: We are resetting the form with whatever is in data.
     * We don't know if this is the best way to do it in React. */
    reset({ ...data });
    onAction(
      // fake target
      {
        name: 'edit.start',
      },
      {
        ...data,
      }
    );

    enterEditMode();
  };

  const onSubmit = (incomingData) => {
    onAction(
      // fake target
      {
        name: 'edit.save',
      },
      {
        ...data,
        ...incomingData,
      }
    );

    exitEditMode();
  };

  return isEditing ? (
    <>
      <Button
        {...cancelTestId}
        className='ui negative icon button'
        disabled={isSubmitting}
        onClick={handleCancel}
        type='button'
      >
        <Icon name='delete' />
      </Button>
      <Button
        {...saveTestId}
        className='ui positive icon button'
        disabled={!isDirty || isSubmitting}
        onClick={handleSubmit(onSubmit)}
        type='button'
      >
        <Icon name='check' />
      </Button>
    </>
  ) : (
    <Button
      {...editTestId}
      className='ui basic button'
      onClick={handleEdit}
      type='button'
    >
      {editLabel}
    </Button>
  );
};

ActionEdit.propTypes = propTypes;

export { propTypes };
export default ActionEdit;
