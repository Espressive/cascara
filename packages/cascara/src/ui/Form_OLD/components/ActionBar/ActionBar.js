import React from 'react';
import pt from 'prop-types';

// external prop types
import { propTypes as actionButtonPT } from '../../../../modules/ActionButton';
import { propTypes as actionEditPT } from '../../modules/ActionEdit';

import styles from '../../Form_OLD.module.scss';
import { Boundaries } from '../../../../system-components';

const propTypes = {
  actions: pt.arrayOf(
    pt.oneOfType([pt.shape(actionButtonPT), pt.shape(actionEditPT)])
  ),
};

const ActionBar = ({ actions }) => {
  return (
    <Boundaries>
      <div className={styles.FormActionBar}>{actions}</div>
    </Boundaries>
  );
};

ActionBar.propTypes = propTypes;

export default ActionBar;
