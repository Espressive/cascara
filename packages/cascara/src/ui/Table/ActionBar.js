import React from 'react';
import pt from 'prop-types';
// import styles from './Table.module.scss';

// external prop types
import { propTypes as actionButtonPT } from '../../modules/ActionButton';
import { propTypes as actionEditPT } from '../../modules/ActionEdit';

const propTypes = {
  actions: pt.arrayOf(
    pt.oneOfType([pt.shape(actionButtonPT), pt.shape(actionEditPT)])
  ),
  title: pt.string,
};

const ActionBar = ({ actions = [], title = '' }) => (
  <>
    {title && <span>{title}</span>}
    {actions.length && actions}
  </>
);

ActionBar.propTypes = propTypes;

export default ActionBar;
