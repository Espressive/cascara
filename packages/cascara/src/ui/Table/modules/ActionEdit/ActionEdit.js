import React, { useCallback, useContext } from 'react';
import pt from 'prop-types';

import { ModuleContext } from '../../../../modules/context';
import { Button } from 'reakit';
import { Icon } from 'semantic-ui-react';

const propTypes = {
  cancelLabel: pt.string,
  dataTestIDs: pt.shape({
    cancel: pt.string,
    edit: pt.string,
    save: pt.string,
  }),
  editLabel: pt.string,
  saveLabel: pt.string,
};

const ActionEdit = ({ dataTestIDs, editLabel = 'Edit' }) => {
  const {
    idOfRecordInEditMode,
    isEditing,
    enterEditMode,
    exitEditMode,
    formMethods,
    record,
    onAction,
    uniqueIdAttribute,
  } = useContext(ModuleContext);
  const { handleSubmit, formState, reset } = formMethods;
  const { isDirty, isSubmitting } = formState;
  const recordId = record[uniqueIdAttribute];
  const whenAnotherRowIsEditing = Boolean(idOfRecordInEditMode);

  // this seems like ugly, we need to find a better way
  // to ease testing..
  const cancelTestId = {};
  const editTestId = {};
  const saveTestId = {};

  if (typeof dataTestIDs === 'object') {
    cancelTestId['data-testid'] = dataTestIDs['cancel'];
    editTestId['data-testid'] = dataTestIDs['edit'];
    saveTestId['data-testid'] = dataTestIDs['save'];
  }

  const handleReset = useCallback(() => {
    onAction(
      // fake target
      {
        name: 'edit.cancel',
      },
      {
        ...record,
      }
    );

    exitEditMode();
  }, [exitEditMode, onAction, record]);

  const handleCancel = useCallback(() => {
    isDirty
      ? // eslint-disable-next-line no-restricted-globals, no-alert -- For now we do not have our own confirmation dialog so we are using native confirms
        confirm('Abandon unsaved changes?') && handleReset()
      : handleReset();
  }, [handleReset, isDirty]);

  const handleEdit = useCallback(() => {
    // FDS-91: We are resetting the form with whatever is in record.
    // We don't know if this is the best way to do it in React.
    reset({ ...record });
    onAction(
      // fake target
      {
        name: 'edit.start',
      },
      {
        ...record,
      }
    );

    enterEditMode(recordId);
  }, [enterEditMode, onAction, record, recordId, reset]);

  const onSubmit = useCallback(
    (data) => {
      onAction(
        // fake target
        {
          name: 'edit.save',
        },
        {
          ...record,
          ...data,
        }
      );

      exitEditMode();
    },
    [exitEditMode, onAction, record]
  );

  return isEditing ? (
    <>
      <Button
        {...cancelTestId}
        className='ui negative icon button'
        disabled={isSubmitting}
        name={'edit.cancel'}
        onClick={handleCancel}
        type='button'
      >
        <Icon name='delete' />
      </Button>
      <Button
        {...saveTestId}
        className='ui positive icon button'
        disabled={!isDirty || isSubmitting}
        name={'edit.save'}
        onClick={handleSubmit(onSubmit)}
        type='button'
      >
        <Icon name='check' />
      </Button>
    </>
  ) : (
    <Button
      {...editTestId}
      className='ui basic button'
      disabled={whenAnotherRowIsEditing}
      name={'edit.start'}
      onClick={handleEdit}
      type='button'
    >
      {editLabel}
    </Button>
  );
};

ActionEdit.propTypes = propTypes;

export { ActionEdit, propTypes };
export default ActionEdit;
