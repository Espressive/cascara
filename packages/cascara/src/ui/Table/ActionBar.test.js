import React from 'react';
import { render } from '@testing-library/react';

import ActionBar from './ActionBar';
import ActionButton from '../../modules/ActionButton';

const ACTION_MODULES = {
  button: ActionButton,
};

describe('ActionBar', () => {
  /**
   *
   */
  describe('component tree', () => {
    const actions = [
      {
        'data-testid': 'new',
        isLabeled: false,
        label: 'new',
        module: 'button',
        name: 'new',
        size: 'small',
      },
      {
        'data-testid': 'edit',
        isLabeled: false,
        label: 'edit',
        module: 'button',
        name: 'edit',
        size: 'small',
      },
    ].map((action) => {
      const { module, ...rest } = action;
      const Action = ACTION_MODULES[module];

      /**
       * In certain predefined-action modules in which a label is not required, e.g. `edit`,
       * the following unique key generation fails, as it relies on the label (content).
       */
      const key = `${module}.${rest.label || module}`;

      return <Action key={key} {...rest} />;
    });
    let view;

    test('snapshot', () => {
      view = render(<ActionBar actions={actions} title={'Espressive'} />)
        .container;

      expect(view).toMatchSnapshot();
    });

    test('default props', () => {
      view = render(<ActionBar />).container;

      expect(view).toMatchSnapshot();
    });
  });
});
