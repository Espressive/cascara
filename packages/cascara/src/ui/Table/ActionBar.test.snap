// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ActionBar component tree default props 1`] = `
<div>
  
  0
</div>
`;

exports[`ActionBar component tree snapshot 1`] = `
<div>
  <span>
    Espressive
  </span>
  <button
    aria-label="new"
    class="ui basic button"
    data-testid="new"
    name="new"
    type="button"
  >
    new
  </button>
  <button
    aria-label="edit"
    class="ui basic button"
    data-testid="edit"
    name="edit"
    type="button"
  >
    edit
  </button>
</div>
`;
